movement/monge_ampere.py

```diff
@@ -164,19 +174,74 @@ def _update_coordinates(self):
             self.grad_phi.interpolate(self._grad_phi)
         self.x.assign(self.xi + self.grad_phi)
         self.mesh.coordinates.assign(self.x)
-
         if hasattr(self, "tangling_checker"):
             self.tangling_checker.check()
 
+    def _l2_projector_bcs(self, boundary_tag):
+        r"""
+        Determine boundary conditions to apply for the :math:`L^2` projection step for
+        a given boundary segment.
+
+        Note that the boundary segment *must* be a straight line or plane.
+
+        :arg boundary_tag: tag for the boundary segment in consideration
+        :type boundary_tag: :class:`str`
+        :returns: tuple of boundary conditions
+        :rtype: :class:`tuple` of :class:`~.DirichletBC`\s
+        """
+        zero_bc = firedrake.DirichletBC(self.P1_vec, 0, boundary_tag)
+        if self.fix_boundary_nodes or self.dim == 1:
+            return (zero_bc,)
+        ds = self.ds(boundary_tag)
+
+        # Check for axis-aligned boundaries
+        n = ufl.FacetNormal(self.mesh)
+        iszero = [np.isclose(firedrake.assemble(abs(ni) * ds), 0.0) for ni in n]
+        nzero = sum(iszero)
+        assert nzero < self.dim
+        if nzero == self.dim - 1:
+            idx = iszero.index(False)
+            return (firedrake.DirichletBC(self.P1_vec.sub(idx), 0, boundary_tag),)
+
+        # Enforce no mesh movement normal to boundaries
+        u_cts = firedrake.TrialFunction(self.P1_vec)
+        v_cts = firedrake.TestFunction(self.P1_vec)
+        a_bc = ufl.dot(v_cts, n) * ufl.dot(u_cts, n) * ds
+        L_bc = ufl.dot(v_cts, n) * firedrake.Constant(0.0) * ds
+        bc1 = firedrake.EquationBC(a_bc == L_bc, self._grad_phi, boundary_tag)
+
+        # Check the current boundary segment is a straight line or plane
+        facet_indices = set(self.mesh.exterior_facets.unique_markers)
+        ffacet_indices = [
+            (tag, boundary_tag) for tag in facet_indices.difference([boundary_tag])
+        ]
+        ffacet_bc = firedrake.DirichletBC(self.P1_vec, 0, ffacet_indices)
```

Comment:
Just for my understanding: `ffacet_bc` says that there is no movement of nodes connecting two different facets? Maybe this could benefit from an explicit comment - and then relating it to tangential movement in `bc2` below :)

Reply:
Good point. Hopefully 8112172 makes it clearer.